# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
# Import your libraries:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split as tts
from sklearn import metrics

from sklearn.linear_model import Ridge, Lasso
from sklearn.linear_model import SGDRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.svm import SVR

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function.

In [2]:
# Your code here:
from sklearn.datasets import load_diabetes
#diabetes = pd.DataFrame(load_diabetes().data, columns = load_diabetes().feature_names)
diabetes = load_diabetes()

Let's explore this variable by looking at the different attributes. Do this by looking at the `keys()` of this variable.

In [3]:
# Your code here:

diabetes.keys()

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename', 'data_module'])

The next step is to read the description of the dataset. Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable

In [4]:
# Your code here:

diabetes.DESCR

'.. _diabetes_dataset:\n\nDiabetes dataset\n----------------\n\nTen baseline variables, age, sex, body mass index, average blood\npressure, and six blood serum measurements were obtained for each of n =\n442 diabetes patients, as well as the response of interest, a\nquantitative measure of disease progression one year after baseline.\n\n**Data Set Characteristics:**\n\n  :Number of Instances: 442\n\n  :Number of Attributes: First 10 columns are numeric predictive values\n\n  :Target: Column 11 is a quantitative measure of disease progression one year after baseline\n\n  :Attribute Information:\n      - age     age in years\n      - sex\n      - bmi     body mass index\n      - bp      average blood pressure\n      - s1      tc, total serum cholesterol\n      - s2      ldl, low-density lipoproteins\n      - s3      hdl, high-density lipoproteins\n      - s4      tch, total cholesterol / HDL\n      - s5      ltg, possibly log of serum triglycerides level\n      - s6      glu, blood sugar

What are the variables in this dataset according to the description? List them in the markdown cell below

#### Enter your answer here:




- age
- sex
- body mass index
- average blood pressure
- six blood serum measurements for each of n = 442 diabetes patients
- the response of interest (a quantitative measure of disease progression one year after baseline)

Now explore the data. Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). In the cell below find the shape of the numpy array contained in the data portion of the diabetes variable.

In [6]:
# Your code here:
df_diabetes = pd.DataFrame(load_diabetes().data, columns = load_diabetes().feature_names)
df_diabetes.shape

(442, 10)

In [7]:
df_diabetes

age       sex       bmi        bp        s1        s2        s3  \
0    0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1   -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2    0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3   -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4    0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
..        ...       ...       ...       ...       ...       ...       ...   
437  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566 -0.028674   
438 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165 -0.028674   
439  0.041708  0.050680 -0.015906  0.017293 -0.037344 -0.013840 -0.024993   
440 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283 -0.028674   
441 -0.045472 -0.044642 -0.073030 -0.081413  0.083740  0.027809  0.173816   

           s4        s5        s6  
0   -0.002592  0.019907 -0.017646  
1   -0.039493 -0.068332 -0.092204  
2   -0.002592  0.002861 -0.025930  
3    0.034309  0.022688 -0.009362  
4   -0.002592 -0.031988 -0.046641  
..        ...       ...       ...  
437 -0.002592  0.031193  0.007207  
438  0.034309 -0.018114  0.044485  
439 -0.011080 -0.046883  0.015491  
440  0.026560  0.044529 -0.025930  
441 -0.039493 -0.004222  0.003064  

[442 rows x 10 columns]

# Challenge 2 - Perform Supervised Learning on the Dataset

#### The data has already been split to predictor and response variables. The response variable is in the `target` portion of the variable. 

Given this information, let's apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset. In the cell below, import the linear regression class from sklearn. 

In [11]:
# Your code here:

from sklearn.linear_model import LinearRegression as LinReg
target = diabetes.target

Initialize the model in the variable `diabetes_model`

In [9]:
# Your code here:

diabetes_model = LinReg()

In the cell below, fit the model and print the intercept and coefficients of the model. 

In [12]:
# Your code here:

diabetes_model.fit(df_diabetes, target)

LinearRegression()

In [13]:
diabetes_model.intercept_

152.13348416289597

In [14]:
diabetes_model.coef_

array([ -10.0098663 , -239.81564367,  519.84592005,  324.3846455 ,
       -792.17563855,  476.73902101,  101.04326794,  177.06323767,
        751.27369956,   67.62669218])

# Bonus Challenge 1 - Conduct a Hypothesis Test on the Model

Once we have generated a linear model, we can test each coefficient using a t-test to see whether the confidence interval for the variable contains zero. We can also perform an overall F test to check whether at least one coefficient is significantly different from zero. 

Refer to the resource in this [link](https://onlinecourses.science.psu.edu/stat501/node/297/) for more details and perform the t-tests for the model above. Additionally, interpret the results and list coefficients are significantly different from zero.


Hint: use the statsmodels package.

Your result should look similar to this:

![ols](../ols-results.png)

In [182]:
# Your code here:



# Challenge 2 - Peform Supervised Learning on a Pandas Dataframe

Now that we have looked at data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [15]:
# Your code here:
auto = pd.read_csv('../auto-mpg.csv')

Look at the first 5 rows using the `head()` function:

In [16]:
# Your code here:
auto.head(5)

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [17]:
# Your code here:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

What is the newest model year and the oldest model year?

In [18]:
# Your code here:
#oldest
auto.model_year.min()

70

In [19]:
#newest
auto.model_year.max()

82

Check the dataset for missing values and remove all rows containing at least one missing value.

In [22]:
# Your code here:

auto.isnull().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [23]:
auto = auto.dropna()

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [24]:
# Your code here:

auto.cylinders.value_counts()

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

In [36]:
X = auto.drop(["mpg", "car_name"], axis=1)

In [37]:
X.sample()

cylinders  displacement  horse_power  weight  acceleration  model_year
298          8         350.0        125.0    3900          17.4          79

In [38]:
y = auto.mpg

In [39]:
y.sample()

161    16.0
Name: mpg, dtype: float64

In [40]:
print(X.shape)
print(y.shape)

(392, 6)
(392,)


In [41]:
# Your code here:
X_train, X_test, y_train, y_test = tts(X, y, test_size = 0.2)

Now we will the dataset that we processed and peform linear regression on this data to predict the mpg for each vehicle. Initialize the model in the cell below.

In [42]:
# Your code here:

auto_model = LinReg()

Next, fit the model in the cell below.

In [43]:
# Your code here:
auto_model.fit(X_train, y_train)

LinearRegression()

In [44]:
auto_model.coef_

array([-0.52834515,  0.00817419, -0.00922125, -0.00622954, -0.0184585 ,
        0.75545746])

In [45]:
auto_model.intercept_

-12.815491772604108

# Challenge 3 - Evaluate the Model

the r squared score of a model tells us how much variation is explained by the model. In a typical dataset, most observations differ from the mean. When we create a model, we are trying to generate an equation that will tell us by how much each observation will differ from the mean. Obviously, the vast majority of models are not perfect. They can only predict some of the variation from the mean but not all of it. We attribute the rest of the difference between the actual value and the mean to random error. We would like random error to explain the as little as possible of the variation. This is why the r squared score is an important metric.

In the next cell, compute the r squared score of the model. Do this by first computing the predicted values and assign them to `y_pred`.

In [193]:
# Import the necessary function:



In [51]:
y_pred = auto_model.predict(X_train)

In [53]:
# Your code here:
print("MAE - Mean absolute error", metrics.mean_absolute_error(y_train, y_pred))
print("MSE - Mean squeared error", metrics.mean_squared_error(y_train, y_pred))
print("RMSE - Root mean squared error", np.sqrt(metrics.mean_absolute_error(y_train, y_pred)))
print("R2 - Mean absolute error", metrics.r2_score(y_train, y_pred))

MAE - Mean absolute error 2.6489105690567243
MSE - Mean squeared error 11.381514278872354
RMSE - Root mean squared error 1.6275474091579405
R2 - Mean absolute error 0.8147987205376928


#### Our next step is to evaluate the model using the test data. We would like to ensure that our model is not overfitting the data. This means that our model will not be able to generalize well outside of the training data.

In the cell below, use the model to generate the predicted values for the training data and assign them to `y_test_pred`. Compute the r squared score for the test data by comparing the oberserved `y_test` data and the predicted `y_test_pred`.

In [54]:
y_test_pred = auto_model.predict(X_test)

In [55]:
# Your code here:
print("MAE - Mean absolute error", metrics.mean_absolute_error(y_test, y_test_pred))
print("MSE - Mean squeared error", metrics.mean_squared_error(y_test, y_test_pred))
print("RMSE - Root mean squared error", np.sqrt(metrics.mean_absolute_error(y_test, y_test_pred)))
print("R2 - Mean absolute error", metrics.r2_score(y_test, y_test_pred))

MAE - Mean absolute error 2.49026946580944
MSE - Mean squeared error 12.649992648402504
RMSE - Root mean squared error 1.578058765005106
R2 - Mean absolute error 0.7819490982070567


# Challenge 4 - Improve the Model Fit

While the most common way to improve the fit of a model is by using regularization, there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [56]:
# Your code here:
X_train09, X_test09, y_train09, y_test09 = tts(X, y, test_size = 0.1)

Initialize a new model. Name this model `auto_model09`. Fit the model to the new sample data.

In [57]:
# Your code here:
auto_model09 = LinReg()

In [58]:
auto_model09.fit(X_train09, y_train09)

LinearRegression()

In [59]:
auto_model.coef_

array([-0.52834515,  0.00817419, -0.00922125, -0.00622954, -0.0184585 ,
        0.75545746])

In [60]:
auto_model.intercept_

-12.815491772604108

Compute the predicted values and r squared score for our new model and new sample data.

In [61]:
y_pred09 = auto_model09.predict(X_train09)

In [62]:
# Your code here:
print("MAE - Mean absolute error", metrics.mean_absolute_error(y_train09, y_pred09))
print("MSE - Mean squeared error", metrics.mean_squared_error(y_train09, y_pred09))
print("RMSE - Root mean squared error", np.sqrt(metrics.mean_absolute_error(y_train09, y_pred09)))
print("R2 - Mean absolute error", metrics.r2_score(y_train09, y_pred09))

MAE - Mean absolute error 2.64997610010619
MSE - Mean squeared error 11.80368570767731
RMSE - Root mean squared error 1.6278747187993892
R2 - Mean absolute error 0.8112462683884708


Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [63]:
y_test_pred09 = auto_model09.predict(X_test09)

In [64]:
# Your code here:
print("MAE - Mean absolute error", metrics.mean_absolute_error(y_test09, y_test_pred09))
print("MSE - Mean squeared error", metrics.mean_squared_error(y_test09, y_test_pred09))
print("RMSE - Root mean squared error", np.sqrt(metrics.mean_absolute_error(y_test09, y_test_pred09)))
print("R2 - Mean absolute error", metrics.r2_score(y_test09, y_test_pred09))

MAE - Mean absolute error 2.5885108753293395
MSE - Mean squeared error 10.306478201746756
RMSE - Root mean squared error 1.608884978899778
R2 - Mean absolute error 0.7550057410764442


# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [200]:
# Your code here:



Fit the model and print the ranking

In [201]:
# Your code here:



Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [202]:
# Your code here:



Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [203]:
# Your code here: 

